File: app/src/main/java/com/kobrakid/retroachievements/adapter/AchievementAdapter.java
package com.kobrakid.retroachievements.adapter;

import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.os.Bundle;
import android.transition.Slide;
import android.transition.TransitionSet;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import com.kobrakid.retroachievements.Consts;
import com.kobrakid.retroachievements.GameDetailsActivity;
import com.kobrakid.retroachievements.R;
import com.kobrakid.retroachievements.fragment.AchievementDetailsFragment;
import com.kobrakid.retroachievements.fragment.AchievementSummaryFragment;
import com.squareup.picasso.Picasso;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Objects;

public class AchievementAdapter extends RecyclerView.Adapter<AchievementAdapter.AchievementViewHolder> {

    private final ArrayList<String> ids;
    private final ArrayList<String> badges;
    private final ArrayList<String> titles;
    private final ArrayList<String> points;
    private final ArrayList<String> trueRatios;
    private final ArrayList<String> descriptions;
    private final ArrayList<String> datesEarned;
    private final ArrayList<String> numsAwarded;
    private final ArrayList<String> numsAwardedHC;
    private final ArrayList<String> authors;
    private final ArrayList<String> datesCreated;
    private final ArrayList<String> datesModified;
    private final ArrayList<Boolean> hardcoreEarnings;
    private double numDistinctCasual = 1;

    private final Fragment fragment;
    private final AchievementViewHolderListener viewHolderListener;

    public AchievementAdapter(Fragment fragment) {
        this.fragment = fragment;
        ids = new ArrayList<>();
        badges = new ArrayList<>();
        titles = new ArrayList<>();
        points = new ArrayList<>();
        trueRatios = new ArrayList<>();
        descriptions = new ArrayList<>();
        datesEarned = new ArrayList<>();
        numsAwarded = new ArrayList<>();
        numsAwardedHC = new ArrayList<>();
        authors = new ArrayList<>();
        datesCreated = new ArrayList<>();
        datesModified = new ArrayList<>();
        hardcoreEarnings = new ArrayList<>();
        viewHolderListener = new AchievementViewHolderListenerImpl(fragment, this);

    }

    @NonNull
    @Override
    public AchievementAdapter.AchievementViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        LinearLayout linearLayout = (LinearLayout) LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.view_holder_achievement_summary,
                        parent,
                        false);
        return new AchievementViewHolder(linearLayout, viewHolderListener);
    }

    @Override
    public void onBindViewHolder(@NonNull AchievementViewHolder holder, int position) {

        // Hidden Text Views
        ((TextView) holder.linearLayout.findViewById(R.id.recycler_view_position)).setText(String.valueOf(position));
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_author)).setText(authors.get(position));
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_created)).setText(datesCreated.get(position));
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_modified)).setText(datesModified.get(position));

        // Badge
        if (hardcoreEarnings.get(position)) {
            (holder.linearLayout.findViewById(R.id.achievement_summary_badge)).setBackground(Objects.requireNonNull(fragment.getContext()).getDrawable(R.drawable.image_view_border));
        } else {
            (holder.linearLayout.findViewById(R.id.achievement_summary_badge)).setBackground(null);
        }
        Picasso.get()
                .load(Consts.BASE_URL + "/" + Consts.GAME_BADGE_POSTFIX + "/" + badges.get(position) + ".png")
                .into((ImageView) holder.linearLayout.findViewById(R.id.achievement_summary_badge));
        holder.linearLayout.findViewById(R.id.achievement_summary_badge).setTransitionName("achievement_" + position);
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_badge_id))
                .setText(badges.get(position));

        // Text descriptions
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_id))
                .setText(ids.get(position));
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_title))
                .setText(fragment.getString(R.string.achievement_summary_title, titles.get(position), points.get(position), trueRatios.get(position)));
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_desc))
                .setText(descriptions.get(position));
        if (datesEarned.get(position).startsWith("NoDate")) {
            ColorMatrix matrix = new ColorMatrix();
            matrix.setSaturation(0);
            ((ImageView) holder.linearLayout.findViewById(R.id.achievement_summary_badge)).setColorFilter(new ColorMatrixColorFilter(matrix));
            ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_date))
                    .setText("");
        } else {
            ((ImageView) holder.linearLayout.findViewById(R.id.achievement_summary_badge)).clearColorFilter();
            ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_date))
                    .setText(Objects.requireNonNull(fragment.getContext()).getString(R.string.date_earned, datesEarned.get(position)));
        }
        ((TextView) holder.linearLayout.findViewById(R.id.achievement_summary_stats))
                .setText(Objects.requireNonNull(fragment.getContext()).getString(R.string.won_by,
                        numsAwarded.get(position),
                        numsAwardedHC.get(position),
                        (int) numDistinctCasual,
                        new DecimalFormat("@@@@")
                                .format(Double.parseDouble(numsAwarded.get(position)) / numDistinctCasual * 100.0)));

        // Double-layered Progress Bar
        ProgressBar progressBar = holder.linearLayout.findViewById(R.id.achievement_summary_progress);
        progressBar.setProgress((int) (Double.parseDouble(numsAwardedHC.get(position)) / numDistinctCasual * 10000.0));
        progressBar.setSecondaryProgress((int) (Double.parseDouble(numsAwarded.get(position)) / numDistinctCasual * 10000.0));
    }

    @Override
    public int getItemCount() {
        return ids.size();
    }

    public void addAchievement(
            int index,
            String id,
            String badge,
            String title,
            String point,
            String trueRatio,
            String description,
            String dateEarned,
            boolean earnedHardcore,
            String numAwarded,
            String numAwardedHC,
            String author,
            String dateCreated,
            String dateModified) {
        ids.add(index, id);
        badges.add(index, badge);
        titles.add(index, title);
        points.add(index, point);
        trueRatios.add(index, trueRatio);
        descriptions.add(index, description);
        datesEarned.add(index, dateEarned);
        hardcoreEarnings.add(index, earnedHardcore);
        numsAwarded.add(index, numAwarded);
        numsAwardedHC.add(index, numAwardedHC);
        authors.add(index, author);
        datesCreated.add(index, dateCreated);
        datesModified.add(index, dateModified);
        notifyItemInserted(index);
    }

    public void setNumDistinctCasual(double n) {
        numDistinctCasual = n;
    }

    public void clear() {
        ids.clear();
        badges.clear();
        titles.clear();
        points.clear();
        trueRatios.clear();
        descriptions.clear();
        hardcoreEarnings.clear();
        datesEarned.clear();
        numsAwarded.clear();
        numsAwardedHC.clear();
        authors.clear();
        datesCreated.clear();
        datesModified.clear();
    }

    /* Inner Classes and Interfaces */

    private interface AchievementViewHolderListener {
        void onItemClicked(View view, int adapterPosition);
    }

    private static class AchievementViewHolderListenerImpl implements AchievementViewHolderListener {

        private final Fragment fragment;
        private final AchievementAdapter adapter;

        AchievementViewHolderListenerImpl(Fragment fragment, AchievementAdapter adapter) {
            this.fragment = fragment;
            this.adapter = adapter;
        }

        @Override
        public void onItemClicked(View view, int adapterPosition) {// Create a new transition
            TransitionSet transitionSet = new TransitionSet();
            transitionSet.setOrdering(TransitionSet.ORDERING_TOGETHER);
            // Get the adapter position of the first child
            int firstChildIndex = Integer.parseInt(
                    ((TextView) Objects.requireNonNull(((AchievementSummaryFragment) fragment)
                            .layoutManager
                            .getChildAt(0))
                            .findViewById(R.id.recycler_view_position))
                            .getText()
                            .toString());
            // FIXME Animation no longer occurring 😢 (but was already pretty janky)
            // Custom logic to slide higher achievements up, lower ones down
            for (int i = 0; i < adapter.getItemCount(); i++) {
                Slide slide = new Slide();
                slide.setDuration(Objects.requireNonNull(fragment.getActivity()).getResources().getInteger(R.integer.animation_duration));
                slide.addTarget(((AchievementSummaryFragment) fragment).layoutManager.getChildAt(i));
                if (i + firstChildIndex < adapterPosition) {
                    slide.setSlideEdge(Gravity.TOP);
                    transitionSet.addTransition(slide);
                } else if (i + firstChildIndex > adapterPosition) {
                    slide.setSlideEdge(Gravity.BOTTOM);
                    transitionSet.addTransition(slide);
                }
            }
            fragment.setExitTransition(transitionSet);
            // Set up the target fragment
            ImageView transitionBadge = view.findViewById(R.id.achievement_summary_badge);
            Fragment detailsFragment = new AchievementDetailsFragment();
            Bundle bundle = new Bundle();
            bundle.putString("Position", "" + adapterPosition);
            bundle.putString("GameID", adapter.ids.get(adapterPosition));
            bundle.putString("ImageIcon", adapter.badges.get(adapterPosition));
            bundle.putString("Title", adapter.titles.get(adapterPosition));
            bundle.putString("Points", adapter.points.get(adapterPosition));
            bundle.putString("TrueRatio", adapter.trueRatios.get(adapterPosition));
            bundle.putString("Description", adapter.descriptions.get(adapterPosition));
            bundle.putString("DateEarned", adapter.datesEarned.get(adapterPosition));
            bundle.putString("NumAwarded", adapter.numsAwarded.get(adapterPosition));
            bundle.putString("NumAwardedHardcore", adapter.numsAwardedHC.get(adapterPosition));
            bundle.putString("Author", adapter.authors.get(adapterPosition));
            bundle.putString("DateCreated", adapter.datesCreated.get(adapterPosition));
            bundle.putString("DateModified", adapter.datesModified.get(adapterPosition));
            bundle.putDouble("NumDistinctPlayersCasual", adapter.numDistinctCasual);
            detailsFragment.setArguments(bundle);
            Objects.requireNonNull(fragment
                    .getActivity())
                    .getSupportFragmentManager()
                    .beginTransaction()
                    .setReorderingAllowed(true)
                    .addSharedElement(transitionBadge, transitionBadge.getTransitionName())
                    .replace(R.id.game_details_frame, detailsFragment, AchievementDetailsFragment.class.getSimpleName())
                    .addToBackStack(null)
                    .commit();
        }
    }

    public static class AchievementViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        final LinearLayout linearLayout;
        private final AchievementViewHolderListener viewHolderListener;

        AchievementViewHolder(LinearLayout linearLayout, AchievementViewHolderListener viewHolderListener) {
            super(linearLayout);
            this.linearLayout = linearLayout;
            this.viewHolderListener = viewHolderListener;
            linearLayout.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            viewHolderListener.onItemClicked(view, getAdapterPosition());
        }
    }

}

File: app/src/main/java/com/kobrakid/retroachievements/adapter/AchievementAdapter.kt
package com.kobrakid.retroachievements.adapter

import android.content.res.Resources
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.ProgressBar
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.core.os.bundleOf
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.RecyclerView
import androidx.transition.TransitionInflater
import com.kobrakid.retroachievements.Consts
import com.kobrakid.retroachievements.R
import com.kobrakid.retroachievements.adapter.AchievementAdapter.AchievementViewHolder
import com.kobrakid.retroachievements.fragment.AchievementDetailsFragment
import com.kobrakid.retroachievements.fragment.AchievementSummaryFragment
import com.kobrakid.retroachievements.ra.Achievement
import com.squareup.picasso.Picasso
import java.text.DecimalFormat

class AchievementAdapter(fragment: Fragment, private val resources: Resources) : RecyclerView.Adapter<AchievementViewHolder>() {

    private val ids = mutableListOf<String>()
    private val badges = mutableListOf<String>()
    private val titles = mutableListOf<String>()
    private val points = mutableListOf<String>()
    private val trueRatios = mutableListOf<String>()
    private val descriptions = mutableListOf<String>()
    private val datesEarned = mutableListOf<String>()
    private val numsAwarded = mutableListOf<String>()
    private val numsAwardedHC = mutableListOf<String>()
    private val authors = mutableListOf<String>()
    private val datesCreated = mutableListOf<String>()
    private val datesModified = mutableListOf<String>()
    private val hardcoreEarnings = mutableListOf<Boolean>()
    private var numDistinctCasual = 1.0
    private val viewHolderListener = AchievementViewHolderListenerImpl(fragment, this)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AchievementViewHolder {
        return AchievementViewHolder(LayoutInflater
                .from(parent.context)
                .inflate(R.layout.view_holder_achievement_summary,
                        parent,
                        false) as ConstraintLayout, viewHolderListener)
    }

    override fun onBindViewHolder(holder: AchievementViewHolder, position: Int) {

        // Hidden Text Views
        holder.layout.findViewById<TextView>(R.id.recycler_view_position).text = position.toString()
        holder.layout.findViewById<TextView>(R.id.achievement_summary_author).text = authors[position]
        holder.layout.findViewById<TextView>(R.id.achievement_summary_created).text = datesCreated[position]
        holder.layout.findViewById<TextView>(R.id.achievement_summary_modified).text = datesModified[position]

        // Badge
        with(holder.layout.findViewById<ImageView>(R.id.achievement_summary_badge)) {
            background = if (hardcoreEarnings[position]) context.getDrawable(R.drawable.image_view_border) else null
            Picasso.get()
                    .load(Consts.BASE_URL + "/" + Consts.GAME_BADGE_POSTFIX + "/" + badges[position] + ".png")
                    .placeholder(R.drawable.favicon)
                    .into(this)
        }
        holder.layout.findViewById<TextView>(R.id.achievement_summary_badge_id).text = badges[position]

        // Text descriptions
        holder.layout.findViewById<TextView>(R.id.achievement_summary_id).text = ids[position]
        holder.layout.findViewById<TextView>(R.id.achievement_summary_title).text = resources.getString(R.string.achievement_summary_title, titles[position], points[position], trueRatios[position])
        holder.layout.findViewById<TextView>(R.id.achievement_summary_desc).text = descriptions[position]
        if (datesEarned[position].startsWith("NoDate")) {
            val matrix = ColorMatrix()
            matrix.setSaturation(0f)
            holder.layout.findViewById<ImageView>(R.id.achievement_summary_badge).colorFilter = ColorMatrixColorFilter(matrix)
            holder.layout.findViewById<TextView>(R.id.achievement_summary_date).text = ""
        } else {
            holder.layout.findViewById<ImageView>(R.id.achievement_summary_badge).clearColorFilter()
            holder.layout.findViewById<TextView>(R.id.achievement_summary_date).text = resources.getString(R.string.date_earned_lower, datesEarned[position])
        }
        holder.layout.findViewById<TextView>(R.id.achievement_summary_stats).text = resources.getString(R.string.won_by,
                numsAwarded[position],
                numsAwardedHC[position],
                numDistinctCasual.toInt(),
                DecimalFormat("@@@@")
                        .format(numsAwarded[position].toDouble() / numDistinctCasual * 100.0))

        // Double-layered Progress Bar
        val progressBar = holder.layout.findViewById<ProgressBar>(R.id.achievement_summary_progress)
        progressBar.progress = (numsAwardedHC[position].toDouble() / numDistinctCasual * 10000.0).toInt()
        progressBar.secondaryProgress = (numsAwarded[position].toDouble() / numDistinctCasual * 10000.0).toInt()
    }

    override fun getItemCount(): Int {
        return ids.size
    }

    fun addAchievement(index: Int, achievement: Achievement) {
        ids.add(index, achievement.id)
        badges.add(index, achievement.badge)
        titles.add(index, achievement.title)
        points.add(index, achievement.point)
        trueRatios.add(index, achievement.trueRatio)
        descriptions.add(index, achievement.description)
        datesEarned.add(index, achievement.dateEarned)
        hardcoreEarnings.add(index, achievement.earnedHardcore)
        numsAwarded.add(index, achievement.numAwarded)
        numsAwardedHC.add(index, achievement.numAwardedHC)
        authors.add(index, achievement.author)
        datesCreated.add(index, achievement.dateCreated)
        datesModified.add(index, achievement.dateModified)
        notifyItemInserted(index)
    }

    fun getAchievementWithId(id: String): Achievement {
        val i = ids.indexOf(id)
        if (i >= 0)
            return Achievement(
                    ids[i],
                    badges[i],
                    titles[i],
                    points[i],
                    trueRatios[i],
                    descriptions[i],
                    datesEarned[i],
                    hardcoreEarnings[i],
                    numsAwarded[i],
                    numsAwardedHC[i],
                    authors[i],
                    datesCreated[i],
                    datesModified[i],
                    numDistinctCasual)
        else
            return Achievement()
    }

    fun setNumDistinctCasual(n: Double) {
        numDistinctCasual = n
    }

    interface AchievementViewHolderListener {
        fun onItemClicked(view: View, adapterPosition: Int)
    }

    inner class AchievementViewHolderListenerImpl internal constructor(private val fragment: Fragment, private val adapter: AchievementAdapter) : AchievementViewHolderListener {
        override fun onItemClicked(view: View, adapterPosition: Int) {
            val transitionBadge = view.findViewById<ImageView>(R.id.achievement_summary_badge)
                    .apply { transitionName = "achievement_$adapterPosition" }
            val sharedTransition = TransitionInflater.from(fragment.context).inflateTransition(R.transition.change_image_transform)
            val transition = TransitionInflater.from(fragment.context).inflateTransition(android.R.transition.slide_bottom)
            fragment.apply {
                sharedElementReturnTransition = sharedTransition
                // FIXME: Has no effect
                exitTransition = transition
            }
            fragment.childFragmentManager
                    .beginTransaction()
                    .setReorderingAllowed(true)
                    .addSharedElement(transitionBadge, transitionBadge.transitionName)
                    .replace(R.id.game_details_frame,
                            AchievementDetailsFragment().apply {
                                arguments = bundleOf(
                                        "achievement" to adapter.getAchievementWithId(adapter.ids[adapterPosition]),
                                        "transitionName" to transitionBadge.transitionName
                                )
                                sharedElementEnterTransition = sharedTransition
                                enterTransition = transition
                            },
                            AchievementDetailsFragment::class.java.simpleName)
                    .addToBackStack(AchievementSummaryFragment::class.java.simpleName)
                    .commit()
        }

    }

    inner class AchievementViewHolder internal constructor(val layout: ConstraintLayout, private val viewHolderListener: AchievementViewHolderListener) : RecyclerView.ViewHolder(layout), View.OnClickListener {
        override fun onClick(view: View) {
            viewHolderListener.onItemClicked(view, adapterPosition)
        }

        init {
            layout.setOnClickListener(this)
        }
    }

}